src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs

```diff
@@ -442,9 +445,24 @@ private void CancelRequestAbortedToken()
         {
             try
             {
-                _abortedCts.Cancel();
-                _abortedCts.Dispose();
-                _abortedCts = null;
+                CancellationTokenSource localAbortCts = null;
+                var shouldCancel = false;
+                lock (_abortLock)
+                {
+                    if (_abortedCts != null && !_preventRequestAbortedCancellation)
+                    {
+                        localAbortCts = _abortedCts;
+                        _abortedCts = null;
+                        shouldCancel = true;
+                    }
+                }
+
+                // If we cancel the cts, we don't dispose as people may still be using
+                // the cts. It also isn't necessary to dispose a canceled cts.
+                if (shouldCancel)
```

Comment:
nit:

```suggestion
                if (localAbortCts != null)
```

Reply:
smart.

Reply:
Or just go back to `localAbortedCts?.Cancel()`.